ot/gmm.py

```diff
@@ -0,0 +1,332 @@
+    of the GMM-OT plan, and is called T_bary in [69]. The 'random' mapping takes
+    for each input point a random pair (i,j) of components of the GMMs and
+    applied the Gaussian map, it is called T_rand in [69].
+
+    Parameters
+    ----------
+    x : array-like, shape (n_samples, d)
+        Input data points.
+
+    m_s : array-like, shape (k_s, d)
+        Mean vectors of the source GMM components.
+
+    m_t : array-like, shape (k_t, d)
+        Mean vectors of the target GMM components.
+
+    C_s : array-like, shape (k_s, d, d)
+        Covariance matrices of the source GMM components.
+
+    C_t : array-like, shape (k_t, d, d)
+        Covariance matrices of the target GMM components.
+
+    w_s : array-like, shape (k_s,)
+        Weights of the source GMM components.
+
+    w_t : array-like, shape (k_t,)
+        Weights of the target GMM components.
+
+    plan : array-like, shape (k_s, k_t), optional
+        Optimal transport plan between the source and target GMM components.
+        If not provided, it will be computed internally.
+
+    method : {'bary', 'rand'}, optional
+        Method for applying the GMM OT mapping. 'bary' uses barycentric mapping,
+        while 'rand' uses random sampling. Default is 'bary'.
+
+    seed : int, optional
+        Seed for the random number generator. Only used when method='rand'.
+
+    Returns
+    -------
+    out : array-like, shape (n_samples, d)
+        Output data points after applying the GMM OT mapping.
+
+    References
+    ----------
+    .. [69] Delon, J., & Desolneux, A. (2020). A Wasserstein-type distance in
+    the space of Gaussian mixture models. SIAM Journal on Imaging Sciences,
+    13(2), 936-970.
+
+    """
+
+    if plan is None:
+        plan = gmm_ot_plan(m_s, m_t, C_s, C_t, w_s, w_t)
+        nx = get_backend(x, m_s, m_t, C_s, C_t, w_s, w_t)
+    else:
+        nx = get_backend(x, m_s, m_t, C_s, C_t, w_s, w_t, plan)
+
+    k_s, k_t = m_s.shape[0], m_t.shape[0]
+    d = m_s.shape[1]
+    n_samples = x.shape[0]
+
+    if method == 'bary':
+        # TODO asserts
+        normalization = gmm_pdf(x, m_s, C_s, w_s)[:, None]
+        out = nx.zeros(x.shape)
+
+        for i in range(k_s):
+            Cs12 = nx.sqrtm(C_s[i])
+            Cs12inv = nx.inv(Cs12)
+
+            for j in range(k_t):
+                g = gaussian_pdf(x, m_s[i], C_s[i])[:, None]
+
+                M0 = nx.sqrtm(Cs12 @ C_t[j] @ Cs12)
+                A = Cs12inv @ M0 @ Cs12inv
+                b = m_t[j] - A @ m_s[i]
+
+                # gaussian mapping between components i and j applied to x
+                T_ij_x = x @ A + b
+                out = out + plan[i, j] * g * T_ij_x
+
+        return out / normalization
+
+    else:  # rand
+        # A[i, j] is the linear part of the gaussian mapping between components
+        # i and j, b[i, j] is the translation part
+        rng = np.random.RandomState(seed)
+
+        A = nx.zeros((k_s, k_t, d, d))
```

Comment:
this is a BIIG tensor!, again mayb can be done with einsum

Reply:
looping over nonzero entries in the plan is probably faster for this function, since there are at most `k_s + k_t - 1` nonzero entries. I'll look into that